File: app/src/main/java/com/nutriscan/ml/FoodMatchingService.kt
package com.nutriscan.ml

import android.util.Log
import com.nutriscan.data.local.entity.FoodItem
import com.nutriscan.data.repository.FoodRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for matching ML classification results against the food database.
 * Handles normalization, indexing, and ranked matching.
 */
@Singleton
class FoodMatchingService @Inject constructor(
    private val foodRepository: FoodRepository
) {
    companion object {
        private const val TAG = "FoodMatchingService"
    }
    
    // Lazy-initialized index, built on first use
    private var aliasIndex: FoodAliasIndex? = null
    
    /**
     * Initialize the alias index from database.
     * Always rebuilds to pick up any DB changes.
     */
    suspend fun initialize() {
        val foods = withContext(Dispatchers.IO) {
            foodRepository.getAllFoods().first()
        }
        
        Log.d(TAG, "Building index from ${foods.size} foods")
        
        // Build the index on a background thread (heavy string normalization)
        aliasIndex = withContext(Dispatchers.Default) {
            FoodAliasIndex(foods)
        }
        
        Log.d(TAG, "Index built: ${aliasIndex!!.size()} names, aliases loaded")
    }
    
    /**
     * Whether the alias index has been built with at least one food item.
     */
    fun isReady(): Boolean = (aliasIndex?.size() ?: 0) > 0
    
    /**
     * Ensure index is ready, initializing if needed.
     */
    private suspend fun ensureIndexReady(): FoodAliasIndex {
        if (aliasIndex == null) {
            initialize()
        }
        return aliasIndex ?: throw IllegalStateException("Failed to initialize food index")
    }
    
    /**
     * Match ML classification results against the food database.
     * 
     * For Food-11 categories (e.g., "Meat", "Bread"), expands to specific
     * foods using Food11CategoryMapper before matching.
     * 
     * @param results Classification results (may be categories or specific foods)
     * @return Ranked list of FoodMatchResult, deduplicated by food
     */
    suspend fun matchClassifications(
        results: List<ClassificationResult>
    ): List<FoodMatchResult> {
        val index = ensureIndexReady()
        
        return results
            .flatMap { result -> 
                // Check if this is a Food-11 category
                val searchTerms = Food11CategoryMapper.getSearchTerms(result.label)
                
                if (searchTerms.size > 1) {
                    // It's a category → expand to specific foods
                    matchCategoryLabel(result.label, result.confidence, searchTerms, index)
                } else {
                    // It's a specific label → match directly
                    matchSingleLabel(result.label, result.confidence, index)
                }
            }
            .sortedByDescending { it.combinedScore }
            .distinctBy { it.matchedFood?.id ?: it.mlLabel }
            .take(10)  // More candidates for categories
    }
    
    /**
     * Match a Food-11 category by searching for expanded terms.
     */
    private fun matchCategoryLabel(
        category: String,
        confidence: Float,
        searchTerms: List<String>,
        index: FoodAliasIndex
    ): List<FoodMatchResult> {
        val matches = mutableListOf<FoodMatchResult>()
        
        for (term in searchTerms) {
            // Try exact name match
            index.findByExactName(term)?.let { food ->
                matches.add(FoodMatchResult(
                    mlLabel = category,  // Keep original category as ML label
                    normalizedLabel = term,  // But use specific term for display
                    confidence = confidence * 0.95f,  // Slight penalty for expansion
                    matchedFood = food,
                    matchType = MatchType.ALIAS  // Category → food is like alias
                ))
            }
            
            // Try alias match
            index.findByAlias(term)?.let { food ->
                if (matches.none { it.matchedFood?.id == food.id }) {
                    matches.add(FoodMatchResult(
                        mlLabel = category,
                        normalizedLabel = term,
                        confidence = confidence * 0.9f,
                        matchedFood = food,
                        matchType = MatchType.ALIAS
                    ))
                }
            }
        }
        
        return if (matches.isNotEmpty()) {
            matches.sortedByDescending { it.combinedScore }
        } else {
            listOf(createNoMatchResult(category, category.lowercase(), confidence))
        }
    }
    
    /**
     * Match a single ML label against the database.
     * Tries matching in order: exact → alias → token → partial
     */
    private fun matchSingleLabel(
        label: String,
        confidence: Float,
        index: FoodAliasIndex
    ): List<FoodMatchResult> {
        val normalized = LabelNormalizer.normalize(label)
        Log.d(TAG, "matchSingleLabel: '$label' → normalized='$normalized'")
        
        // Skip empty or very short labels
        if (normalized.length < 2) {
            Log.d(TAG, "  Skipped (too short)")
            return listOf(createNoMatchResult(label, normalized, confidence))
        }
        
        val tokens = LabelNormalizer.extractTokens(normalized)
        Log.d(TAG, "  Tokens: $tokens")
        val matches = mutableListOf<FoodMatchResult>()
        
        // Try each token for matching
        for (token in tokens) {
            // 1. Exact name match (highest priority)
            val exactMatch = index.findByExactName(token)
            Log.d(TAG, "  Exact '$token' → ${exactMatch?.name ?: "null"}")
            if (exactMatch != null) {
                return listOf(FoodMatchResult(
                    mlLabel = label,
                    normalizedLabel = normalized,
                    confidence = confidence,
                    matchedFood = exactMatch,
                    matchType = if (token == normalized) MatchType.EXACT else MatchType.TOKEN
                ))
            }
            
            // 2. Alias match
            val aliasMatch = index.findByAlias(token)
            Log.d(TAG, "  Alias '$token' → ${aliasMatch?.name ?: "null"}")
            if (aliasMatch != null) {
                matches.add(FoodMatchResult(
                    mlLabel = label,
                    normalizedLabel = normalized,
                    confidence = confidence,
                    matchedFood = aliasMatch,
                    matchType = MatchType.ALIAS
                ))
            }
        }
        
        // Return alias matches if found
        if (matches.isNotEmpty()) {
            Log.d(TAG, "  → Found ${matches.size} alias match(es)")
            return matches
        }
        
        // 3. Fallback: partial match (never auto-selected)
        val partialMatch = index.findByPartialName(normalized)
        Log.d(TAG, "  Partial '$normalized' → ${partialMatch?.name ?: "null"}")
        if (partialMatch != null) {
            return listOf(FoodMatchResult(
                mlLabel = label,
                normalizedLabel = normalized,
                confidence = confidence,
                matchedFood = partialMatch,
                matchType = MatchType.PARTIAL
            ))
        }
        
        // 4. No match found
        Log.d(TAG, "  → NO MATCH for '$label'")
        return listOf(createNoMatchResult(label, normalized, confidence))
    }
    
    private fun createNoMatchResult(
        label: String,
        normalized: String,
        confidence: Float
    ) = FoodMatchResult(
        mlLabel = label,
        normalizedLabel = normalized,
        confidence = confidence,
        matchedFood = null,
        matchType = MatchType.NONE
    )
    
    /**
     * Get the best auto-selectable match from results.
     * Returns null if no match is safe for auto-selection.
     */
    fun getBestAutoSelectMatch(results: List<FoodMatchResult>): FoodMatchResult? {
        return results.firstOrNull { it.isSafeForAutoSelect }
    }
    
    /**
     * Get all valid candidates for user selection.
     */
    fun getValidCandidates(results: List<FoodMatchResult>): List<FoodMatchResult> {
        return results.filter { it.isValidCandidate }
    }
}

File: web/backend/src/services/foodMatchingService.js
/**
 * Food Matching Service — Direct port of FoodMatchingService.kt + Food11CategoryMapper.kt
 * Matches ML classification results against the food database.
 * 4-strategy ranked matching: exact → alias → token → partial.
 */

const { normalize, extractTokens } = require('./labelNormalizer');

// Match type scores (higher = more trustworthy)
const MatchType = {
  EXACT: { name: 'EXACT', score: 100 },
  ALIAS: { name: 'ALIAS', score: 80 },
  TOKEN: { name: 'TOKEN', score: 60 },
  PARTIAL: { name: 'PARTIAL', score: 30 },
  NONE: { name: 'NONE', score: 0 }
};

// Food-11 category → specific food expansion (from Food11CategoryMapper.kt)
const CATEGORY_TO_FOODS = {
  'Bread': ['bread', 'toast', 'bagel', 'croissant', 'baguette', 'roll', 'bun', 'muffin', 'pretzel'],
  'Dairy product': ['milk', 'cheese', 'yogurt', 'butter', 'cream', 'ice cream', 'cottage cheese', 'mozzarella'],
  'Dessert': ['cake', 'pie', 'cookie', 'brownie', 'donut', 'doughnut', 'pastry', 'cupcake', 'cheesecake', 'pudding', 'chocolate'],
  'Egg': ['egg', 'eggs', 'omelette', 'omelet', 'scrambled eggs', 'fried egg', 'boiled egg', 'poached egg'],
  'Fried food': ['french fries', 'fries', 'fried chicken', 'chicken nuggets', 'onion rings', 'tempura', 'fried', 'crispy'],
  'Meat': ['beef', 'steak', 'pork', 'lamb', 'chicken', 'turkey', 'sausage', 'bacon', 'ham', 'burger', 'hamburger', 'meatball'],
  'Noodles-Pasta': ['pasta', 'spaghetti', 'noodles', 'ramen', 'udon', 'pho', 'lasagna', 'macaroni', 'fettuccine', 'penne'],
  'Rice': ['rice', 'fried rice', 'risotto', 'sushi', 'paella', 'biryani', 'pilaf'],
  'Seafood': ['fish', 'salmon', 'tuna', 'shrimp', 'prawn', 'lobster', 'crab', 'oyster', 'squid', 'clam', 'mussel'],
  'Soup': ['soup', 'broth', 'stew', 'chowder', 'bisque', 'pho', 'ramen', 'miso'],
  'Vegetable-Fruit': ['apple', 'banana', 'orange', 'grape', 'strawberry', 'carrot', 'broccoli', 'tomato', 'salad', 'vegetable', 'fruit']
};

/**
 * Get search terms for a Food-11 category.
 */
function getCategorySearchTerms(category) {
  // Exact match
  if (CATEGORY_TO_FOODS[category]) return CATEGORY_TO_FOODS[category];

  // Normalized match
  const trimmed = category.trim();
  for (const [key, terms] of Object.entries(CATEGORY_TO_FOODS)) {
    if (key.toLowerCase() === trimmed.toLowerCase()) return terms;
    if (key.replace('-', '/').toLowerCase() === trimmed.replace('-', '/').toLowerCase()) return terms;
  }

  return [category.toLowerCase()];
}

function isKnownCategory(category) {
  return getCategorySearchTerms(category).length > 1;
}

/**
 * Create a FoodMatchResult object.
 */
function createMatchResult(mlLabel, normalizedLabel, confidence, matchedFood, matchType) {
  const combinedScore = matchedFood
    ? confidence * (matchType.score / 100)
    : 0;

  return {
    mlLabel,
    normalizedLabel,
    confidence,
    matchedFood,
    matchType: matchType.name,
    matchTypeScore: matchType.score,
    combinedScore,
    confidencePercent: Math.round(confidence * 100),
    isValidCandidate: matchedFood != null && matchType !== MatchType.NONE,
    isSafeForAutoSelect: matchedFood != null &&
      (matchType === MatchType.EXACT || matchType === MatchType.ALIAS) &&
      confidence >= 0.5
  };
}

/**
 * Match a single ML label against the database.
 * Tries matching in order: exact → alias → token → partial
 */
function matchSingleLabel(label, confidence, index) {
  const normalized = normalize(label);
  if (normalized.length < 2) {
    return [createMatchResult(label, normalized, confidence, null, MatchType.NONE)];
  }

  const tokens = extractTokens(normalized);
  const matches = [];

  for (const token of tokens) {
    // 1. Exact name match (highest priority)
    const exactMatch = index.findByExactName(token);
    if (exactMatch) {
      return [createMatchResult(
        label, normalized, confidence, exactMatch,
        token === normalized ? MatchType.EXACT : MatchType.TOKEN
      )];
    }

    // 2. Alias match
    const aliasMatch = index.findByAlias(token);
    if (aliasMatch) {
      matches.push(createMatchResult(label, normalized, confidence, aliasMatch, MatchType.ALIAS));
    }
  }

  if (matches.length > 0) return matches;

  // 3. Fallback: partial match
  const partialMatch = index.findByPartialName(normalized);
  if (partialMatch) {
    return [createMatchResult(label, normalized, confidence, partialMatch, MatchType.PARTIAL)];
  }

  // 4. No match
  return [createMatchResult(label, normalized, confidence, null, MatchType.NONE)];
}

/**
 * Match a Food-11 category by searching for expanded terms.
 */
function matchCategoryLabel(category, confidence, searchTerms, index) {
  const matches = [];
  const seenIds = new Set();

  for (const term of searchTerms) {
    const exactMatch = index.findByExactName(term);
    if (exactMatch && !seenIds.has(exactMatch.name)) {
      seenIds.add(exactMatch.name);
      matches.push(createMatchResult(category, term, confidence * 0.95, exactMatch, MatchType.ALIAS));
    }

    const aliasMatch = index.findByAlias(term);
    if (aliasMatch && !seenIds.has(aliasMatch.name)) {
      seenIds.add(aliasMatch.name);
      matches.push(createMatchResult(category, term, confidence * 0.9, aliasMatch, MatchType.ALIAS));
    }
  }

  if (matches.length > 0) {
    return matches.sort((a, b) => b.combinedScore - a.combinedScore);
  }

  return [createMatchResult(category, category.toLowerCase(), confidence, null, MatchType.NONE)];
}

/**
 * Match ML classification results against the food database.
 * @param {Array} results - [{label, confidence, index}]
 * @param {FoodAliasIndex} foodIndex
 * @returns {Array} Ranked FoodMatchResults
 */
function matchClassifications(results, foodIndex) {
  const allMatches = results.flatMap(result => {
    const searchTerms = getCategorySearchTerms(result.label);

    if (searchTerms.length > 1 && isKnownCategory(result.label)) {
      return matchCategoryLabel(result.label, result.confidence, searchTerms, foodIndex);
    } else {
      return matchSingleLabel(result.label, result.confidence, foodIndex);
    }
  });

  // Sort by combined score, deduplicate by food name
  const seen = new Set();
  return allMatches
    .sort((a, b) => b.combinedScore - a.combinedScore)
    .filter(m => {
      const key = m.matchedFood ? m.matchedFood.name : m.mlLabel;
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    })
    .slice(0, 10);
}

function getBestAutoSelectMatch(results) {
  return results.find(r => r.isSafeForAutoSelect) || null;
}

function getValidCandidates(results) {
  return results.filter(r => r.isValidCandidate);
}

module.exports = {
  matchClassifications,
  getBestAutoSelectMatch,
  getValidCandidates,
  MatchType,
  getCategorySearchTerms,
  isKnownCategory
};
